Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,122 @@
+
+import NIO
+
+/// Data was unexpectedly written or read before the SOCKS proxy
+/// connection has been fully established.
+public struct ProxyNotEstablished: Error {
+    
+}
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline.
+public class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    public let supportedAuthenticationMethods: [AuthenticationMethod]
+    public let targetAddress: AddressType
+    public let targetPort: UInt16
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer
+    
+    public init(supportedAuthenticationMethods: [AuthenticationMethod], targetAddress: AddressType, targetPort: UInt16) {
+        precondition(supportedAuthenticationMethods.count > 0,
+                     "At least one supported authentication method required.")
+        precondition(supportedAuthenticationMethods.count <= 255,
+                     "There can't be more than 255 authentication methods listed.")
+        self.supportedAuthenticationMethods = supportedAuthenticationMethods
+        self.state = ClientStateMachine()
+        self.buffered = ByteBuffer()
+        self.targetAddress = targetAddress
+        self.targetPort = targetPort
+    }
+    
+    public func channelInactive(context: ChannelHandlerContext) {
+        context.fireChannelInactive()
+    }
+    
+    public func channelActive(context: ChannelHandlerContext) {
+        let greeting = ClientGreeting(
+            methods: self.supportedAuthenticationMethods
+        )
+        let capacity = 2 + self.supportedAuthenticationMethods.count
+        var buffer = context.channel.allocator.buffer(capacity: capacity)
+        buffer.writeClientGreeting(greeting)
+        self.state.sendClientGreeting(greeting)
+        
+        context.writeAndFlush(self.wrapOutboundOut(buffer), promise: nil)
+        context.fireChannelActive()
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        
+        // if we've established the connection then forward on the data
+        guard !self.state.proxyEstablished else {
+            context.fireChannelRead(data)
+            return
+        }
+        
+        var buffer = self.unwrapInboundIn(data)
+        self.buffered.writeBuffer(&buffer)
+        let save = self.buffered
+        do {
+            guard let action = try self.state.receiveBuffer(&self.buffered) else {
+                self.buffered = save
+                return
+            }
+            self.handleAction(action, context: context)
+        } catch {
+            context.fireErrorCaught(error)
+            context.close(mode: .all, promise: nil)
+        }
```

Comment:
Why do we pass the buffer as an `inout`? The state machine should not modify the data. Also I think the state machine, should not throw. In case of an error the state machine should return an Action like `.closeConnectionAndFireError(error)`

Reply:
I have no problem with the state machine throwing.

Reply:
It makes sense to pass the buffer as `inout` as our state machine is also the parser. If state machine is able to progress, then some data is consumed, otherwise the data is left alone.

Reply:
@fabianfett are you swayed?